### PR TITLE
[CoordinatedGraphics] Use CoordinatedPlatformLayer in the composition

### PR DESCRIPTION
**Pull request by carlosgcampos**
#### cdbcc94d4406d5c42901ffb9fbc67c795ca8b3ce
<pre>
[CoordinatedGraphics] Use CoordinatedPlatformLayer in the composition
<a href="https://bugs.webkit.org/show_bug.cgi?id=284742">https://bugs.webkit.org/show_bug.cgi?id=284742</a>

Reviewed by Alejandro G. Castro.

Update the composition code to use CoordinatedPlatformLayer:

  - CoordinatedPlatformLayer now has a target which is the TextureMapperLayer
    object that will be used by the TextureMapper during the composition.

  - CoordinatedPlatformLayer::flushCompositingState() checks all pending
    changes and writes the changed properties on the target TextureMapperLayer
    and updates the layer contents. CoordinatedGraphicsScene::updateSceneState()
    now only needs to a iterate the layers calling flushCompositingState().

  - A new class CoordinatedSceneState has been added to keep the scene
    state, but instead of passing the state on every composition
    request, the CoordinatedSceneState object is created by the
    LayerTreeHost on construction and passed to the
    CoordinatedGraphicsScene constructor too. It works similar to a
    GraphicsLayer, whenever the scene changes because layers are added
    or removed, a pending state is kept. Then when compositing the
    pending state is committed when layers are asked causing any removed
    layer to be destroyed too.

  - In LayerTreeHost the root layer no longer needs to be a GraphicsLayer,
    and now it&apos;s a CoordinatedPlatformLayer that is used as the root
    layer in the composition too.

* .github/CODEOWNERS:
* Source/WebCore/page/scrolling/coordinated/ScrollingTreeCoordinated.h:
* Source/WebCore/platform/SourcesNicosia.txt: Removed.
* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/nicosia/NicosiaCompositionLayer.h: Removed.
* Source/WebCore/platform/graphics/nicosia/NicosiaPlatformLayer.h: Removed.
* Source/WebCore/platform/graphics/nicosia/NicosiaScene.cpp: Removed.
* Source/WebCore/platform/graphics/nicosia/NicosiaScene.h: Removed.
* Source/WebCore/platform/graphics/texmap/TextureMapperGCGLPlatformLayer.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.cpp:
(WebCore::TextureMapperPlatformLayerProxy::~TextureMapperPlatformLayerProxy):
(WebCore::TextureMapperPlatformLayerProxy::isActiveLocked const):
(WebCore::TextureMapperPlatformLayerProxy::activateOnCompositingThread):
(WebCore::TextureMapperPlatformLayerProxy::invalidate):
(WebCore::TextureMapperPlatformLayerProxy::swapBuffer):
(WebCore::TextureMapperPlatformLayerProxy::pushNextBuffer):
(WebCore::TextureMapperPlatformLayerProxy::dropCurrentBufferWhilePreservingTexture):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStore::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::CoordinatedPlatformLayer):
(WebCore::CoordinatedPlatformLayer::ensureTarget):
(WebCore::CoordinatedPlatformLayer::target const):
(WebCore::CoordinatedPlatformLayer::invalidateTarget):
(WebCore::CoordinatedPlatformLayer::setPosition):
(WebCore::CoordinatedPlatformLayer::setPositionForScrolling):
(WebCore::CoordinatedPlatformLayer::setBoundsOrigin):
(WebCore::CoordinatedPlatformLayer::setBoundsOriginForScrolling):
(WebCore::CoordinatedPlatformLayer::setAnchorPoint):
(WebCore::CoordinatedPlatformLayer::setSize):
(WebCore::CoordinatedPlatformLayer::setTransform):
(WebCore::CoordinatedPlatformLayer::setChildrenTransform):
(WebCore::CoordinatedPlatformLayer::setScrollingNodeID):
(WebCore::CoordinatedPlatformLayer::setDrawsContent):
(WebCore::CoordinatedPlatformLayer::setMasksToBounds):
(WebCore::CoordinatedPlatformLayer::setPreserves3D):
(WebCore::CoordinatedPlatformLayer::setBackfaceVisibility):
(WebCore::CoordinatedPlatformLayer::setOpacity):
(WebCore::CoordinatedPlatformLayer::setContentsVisible):
(WebCore::CoordinatedPlatformLayer::setContentsOpaque):
(WebCore::CoordinatedPlatformLayer::setContentsRect):
(WebCore::CoordinatedPlatformLayer::setContentsRectClipsDescendants):
(WebCore::CoordinatedPlatformLayer::setContentsClippingRect):
(WebCore::CoordinatedPlatformLayer::setContentsScale):
(WebCore::CoordinatedPlatformLayer::setContentsBuffer):
(WebCore::CoordinatedPlatformLayer::setContentsImage):
(WebCore::CoordinatedPlatformLayer::setContentsColor):
(WebCore::CoordinatedPlatformLayer::setContentsTileSize):
(WebCore::CoordinatedPlatformLayer::setContentsTilePhase):
(WebCore::CoordinatedPlatformLayer::setDirtyRegion):
(WebCore::CoordinatedPlatformLayer::setDamage):
(WebCore::CoordinatedPlatformLayer::setFilters):
(WebCore::CoordinatedPlatformLayer::setMask):
(WebCore::CoordinatedPlatformLayer::setReplica):
(WebCore::CoordinatedPlatformLayer::setBackdrop):
(WebCore::CoordinatedPlatformLayer::setBackdropRect):
(WebCore::CoordinatedPlatformLayer::setAnimations):
(WebCore::CoordinatedPlatformLayer::setChildren):
(WebCore::CoordinatedPlatformLayer::setEventRegion):
(WebCore::CoordinatedPlatformLayer::setDebugBorder):
(WebCore::CoordinatedPlatformLayer::setShowRepaintCounter):
(WebCore::CoordinatedPlatformLayer::needsBackingStore const):
(WebCore::CoordinatedPlatformLayer::updateBackingStore):
(WebCore::CoordinatedPlatformLayer::updateContents):
(WebCore::CoordinatedPlatformLayer::purgeBackingStores):
(WebCore::CoordinatedPlatformLayer::flushCompositingState):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.h:
(WebCore::CoordinatedPlatformLayer::id const):
(WebCore::CoordinatedPlatformLayer::compositionLayer const): Deleted.
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
* Source/WebKit/WebProcess/GPU/media/gstreamer/VideoLayerRemoteGStreamer.cpp:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::CoordinatedGraphicsScene::CoordinatedGraphicsScene):
(WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext):
(WebKit::CoordinatedGraphicsScene::updateSceneState):
(WebKit::CoordinatedGraphicsScene::purgeGLResources):
(WebKit::CoordinatedGraphicsScene::applyStateChanges): Deleted.
(WebKit::CoordinatedGraphicsScene::onNewBufferAvailable): Deleted.
(WebKit::texmapLayer): Deleted.
(WebKit::updateBackingStore): Deleted.
(WebKit::CoordinatedGraphicsScene::removeLayer): Deleted.
(WebKit::CoordinatedGraphicsScene::commitSceneState): Deleted.
(WebKit::CoordinatedGraphicsScene::ensureRootLayer): Deleted.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.h:
(WebKit::CoordinatedGraphicsScene::state const):
(WebKit::CoordinatedGraphicsScene::rootLayer): Deleted.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedSceneState.cpp: Added.
(WebKit::CoordinatedSceneState::CoordinatedSceneState):
(WebKit::CoordinatedSceneState::~CoordinatedSceneState):
(WebKit::CoordinatedSceneState::setRootLayerChildren):
(WebKit::CoordinatedSceneState::addLayer):
(WebKit::CoordinatedSceneState::removeLayer):
(WebKit::CoordinatedSceneState::flush):
(WebKit::CoordinatedSceneState::committedLayers):
(WebKit::CoordinatedSceneState::invalidateCommittedLayers):
(WebKit::CoordinatedSceneState::invalidate):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedSceneState.h: Added.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::LayerTreeHost):
(WebKit::LayerTreeHost::~LayerTreeHost):
(WebKit::LayerTreeHost::flushLayers):
(WebKit::LayerTreeHost::updateRootLayer):
(WebKit::LayerTreeHost::setRootCompositingLayer):
(WebKit::LayerTreeHost::setViewOverlayRootLayer):
(WebKit::LayerTreeHost::sizeDidChange):
(WebKit::LayerTreeHost::attachLayer):
(WebKit::LayerTreeHost::detachLayer):
(WebKit::LayerTreeHost::commitSceneState):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::invalidate):
(WebKit::ThreadedCompositor::renderLayerTree):
(WebKit::ThreadedCompositor::requestComposition):
(WebKit::ThreadedCompositor::updateSceneWithoutRendering): Deleted.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsPlayStation.cmake:
* Source/cmake/OptionsWPE.cmake:
* Tools/Scripts/webkit-filter-log:
* Tools/Scripts/webkitpy/port/gtk.py:
(GtkPort.setup_environ_for_server):
* Tools/Scripts/webkitpy/port/wpe.py:
(WPEPort.setup_environ_for_server):

Canonical link: <a href="https://commits.webkit.org/288158@main">https://commits.webkit.org/288158@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/8921e50e13639140dc3ce2f7ab13cafedea79e00

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81633 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1158 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35582 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86172 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32623 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83739 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1189 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8976 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63686 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21414 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84703 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/842 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74285 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43976 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/81086 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/742 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28463 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31078 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74612 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72159 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29059 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87612 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80687 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8870 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6283 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72015 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9053 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70111 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71250 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15329 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14244 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/146 "Failed to checkout and rebase branch from PR 38008") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12710 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8822 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14350 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103098 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8662 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25041 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-collect-continuously (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12183 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10471 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.h [hunk 1]
```diff
@@ -81,14 +75,13 @@ class TextureMapperPlatformLayerProxy final : public ThreadSafeRefCounted<Textur
     void compositorThreadUpdateTimerFired();
 
     Lock m_lock;
-    Compositor* m_compositor { nullptr };
+    CoordinatedPlatformLayer* m_targetLayer { nullptr };
```
Comment (csaavedra) [inline Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.h]: Can this be a smart pointer instead of a raw one?

**Review by csaavedra — COMMENTED** [review]
Not really my turf, but just a minor comment.

File: Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp [hunk 2]
```diff
@@ -383,8 +401,8 @@ void CoordinatedPlatformLayer::setContentsOpaque(bool contentsOpaque)
         return;
 
     m_contentsOpaque = contentsOpaque;
-    m_nicosia.delta.flagsChanged = true;
-    // FIXME: request a full repaint.
+    m_pendingChanges.add(Change::ContentsOpaque);
+    // FIXME: request a full repaint?
```
Comment (alexgcastro) [inline Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:405]: This is scary, maybe add a bug directly. Probably we need a full repaint request, even if it is a performance penalty.

File: Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp [hunk 3]
```diff
@@ -874,6 +787,147 @@ Ref<CoordinatedTileBuffer> CoordinatedPlatformLayer::paint(const IntRect& dirtyR
...
+        for (auto tileID : update.tilesToCreate())
+            m_backingStore->createTile(tileID);
+        for (auto tileID : update.tilesToRemove())
+            m_backingStore->removeTile(tileID);
+        for (const auto& tileUpdate : update.tilesToUpdate())
+            m_backingStore->updateTile(tileUpdate.tileID, tileUpdate.dirtyRect, tileUpdate.tileRect, tileUpdate.buffer.copyRef(), { });
+    } else {
+        layer.setBackingStore(nullptr);
+        layer.setAnimatedBackingStoreClient(nullptr);
+        m_backingStore = nullptr;
```
Comment (alexgcastro) [inline Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:918]: I find the m_backingStore and the m_backingStoreProxy confusing when we set to nullptr. Maybe we can use some function to make sure we encapsulate the update logic of the values? I understand it is dependent but it is in multiple functions.

**Review by alexgcastro — APPROVED** [review]
LGTM! This reduces the code and simplifies the way we understand it, it can reduce the amount of locking we do. I've added a couple of suggestions, maybe for later. Also maybe in the invalidate of the committed layers we could add a flag so that it invalidates itself when requested, to improve performance, but again it can be done later if we check profiled.

File: Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp (hunk 2 above)
Comment (carlosgcampos) [reply Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:405]: We used to do that, but I'm not sure it's needed, maybe setNeedsDisplay is already called when ContentsOpaque changes. So, I removed it on purpose to see if anything fails without it.

File: Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp (hunk 3 above)
Comment (carlosgcampos) [reply Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:918]: It's indeed confusing, one is used in the composition and should be reset in the compositing thread, the other is the proxy that should be reset in the main thread. I think we should think of new names after all refactorings are done.

**Pull request merged**